File: problems/leetcode/0000/53.cpp
// 53. Maximum Subarray
// https://leetcode.com/problems/maximum-subarray/

using namespace std;

#include <vector>

class Solution {
public:
  int maxSubArray(vector<int> &nums) {
    int r = nums[0];

    for (int i = 1, imax = r; i < nums.size(); i++) {
      imax = max(nums[i], imax + nums[i]);
      r = max(r, imax);
    }

    return r;
  }
};

File: problems/leetcode/0000/53/53.py
# 53. Maximum Subarray
# https://leetcode.com/problems/maximum-subarray/description/

from typing import *


class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        n = len(nums)
        globalMax = localMax = nums[0]

        for i in range(1, n):
            localMax = max(nums[i], nums[i] + localMax)
            globalMax = max(globalMax, localMax)

        return globalMax
